# IMPORTS

In [1]:
import numpy as np
import pandas as pd

#
# CREATE DATAFRAME

In [2]:
#Creating dataframe
movies_df=pd.read_csv("kaggle TMDB 5000 movie dataset/tmdb_5000_movies.csv")
credits_df=pd.read_csv("kaggle TMDB 5000 movie dataset/tmdb_5000_credits.csv")

In [3]:
movies_df.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [4]:
credits_df.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [5]:
credits_df.head()["cast"]

0    [{"cast_id": 242, "character": "Jake Sully", "...
1    [{"cast_id": 4, "character": "Captain Jack Spa...
2    [{"cast_id": 1, "character": "James Bond", "cr...
3    [{"cast_id": 2, "character": "Bruce Wayne / Ba...
4    [{"cast_id": 5, "character": "John Carter", "c...
Name: cast, dtype: object

In [6]:
movies_df.head(3)["original_title"].values

array(['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre'],
      dtype=object)

#
# MERGE BOTH DATAFRAME TO SINGLE ONE

In [7]:
df=movies_df.merge(credits_df,on="title")

In [8]:
df.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

#
# ANALYZING ONLY REQURIED COLUMNS AND DROPPING OTHERS

In [9]:
# all the unlisted here are dropped automatically
df=df[["id","title","genres","overview","keywords","cast","crew"]]

In [10]:
df.head(1)

id   title                                             genres  \
0  19995  Avatar  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

#
# PREPROCESSING PART

### Step-1:checking if any missing data in any column and dropping them is any

In [11]:
df.isnull()

id  title  genres  overview  keywords   cast   crew
0     False  False   False     False     False  False  False
1     False  False   False     False     False  False  False
2     False  False   False     False     False  False  False
3     False  False   False     False     False  False  False
4     False  False   False     False     False  False  False
...     ...    ...     ...       ...       ...    ...    ...
4804  False  False   False     False     False  False  False
4805  False  False   False     False     False  False  False
4806  False  False   False     False     False  False  False
4807  False  False   False     False     False  False  False
4808  False  False   False     False     False  False  False

[4809 rows x 7 columns]

In [12]:
df.isnull().sum()

id          0
title       0
genres      0
overview    3
keywords    0
cast        0
crew        0
dtype: int64

In [13]:
df.dropna(inplace=True)

In [14]:
df.isnull().sum()

id          0
title       0
genres      0
overview    0
keywords    0
cast        0
crew        0
dtype: int64

### Step-2:checking if any data of rows of any columns are duplicate ones and dropping if any

In [15]:
df.duplicated().sum()

np.int64(0)

### Step-3:decorating the weired data available in the genres column accordingly to ours need

In [16]:
#viewing what the data is like initially of any one of genre column
df.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [17]:
#from the data print of above we only want to make it like ['Action','Adventure','Fantasy','Science Fiction']
#so to do soo  we can use helper function,

#### But when use our helper function to make the data in the list,got error
###### def converted_to_list(value):
######     L=[]
######     for i in value:
######          L.append(i["name"])
######     return L
###### 
###### converted_to_list('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')
######
######
###### ---------------------------------------------------------------------------
###### TypeError                                 Traceback (most recent call last)
###### Cell In[23], line 1
###### ----> 1 converted_to_list('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')
###### 
###### Cell In[22], line 6, in converted_to_list(value)
######       4 L=[]
######       5 for i in value:
###### ----> 6  L.append(i["name"])
######       7 return L
###### TypeError: string indices must be integers, not 'str'
######
#### this means we have to convert this string of list to list

In [18]:
#To solve this problem we have to import the module called ast(i.e import ast) from which we gonna use it's function called literal_eval()
import ast
ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [19]:
def convert_to_list(value):
    L=[]
    for i in ast.literal_eval(value):
        L.append(i["name"])
    return L

In [20]:
#applying that helper function to our genre column which then updated to new one as we expected
df["genres"]=df["genres"].apply(convert_to_list)

In [21]:
df["genres"]

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [22]:
df.head(1)

id   title                                         genres  \
0  19995  Avatar  [Action, Adventure, Fantasy, Science Fiction]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

### Step-4:decorating the weired data available in the keywords column accordingly to ours need(FOLLOW SAME STEPS AS ABOVE 3)

In [23]:
# def convert_to_list(obj):
#     L=[]
#     for i in ast.literal_eval(obj):
#         L.append(i["name"])
#     return L

# since have the same line of code as in the previous made function(Therefore,not needed)

In [24]:
#applying that helper function to our keywords column which then updated to new one as we expected
df["keywords"]=df["keywords"].apply(convert_to_list)

In [25]:
df.head(1)

id   title                                         genres  \
0  19995  Avatar  [Action, Adventure, Fantasy, Science Fiction]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

### Step-5:decorating the weired data available in the cast column accordingly to ours need(FOLLOW SIMILAR STEPS AS ABOVE)
### HERE,BASICALLY WE WILL BE ONLY GRABBING THE TOP-3 ACTORS.

In [26]:
#viewing what the data is like initially of any one of cast column
df.iloc[0].cast

# or df["cast"][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

In [27]:
def convert_cast_list(value):
    L=[]
    count=0
    for i in ast.literal_eval(value):
        if count !=3 :
            L.append(i["name"])
            count+=1
        else:
            break
    return L

In [28]:
#applying that helper function to our cast column which then updated to new one as we expected
df["cast"]=df["cast"].apply(convert_cast_list)

In [29]:
df.head(1)

id   title                                         genres  \
0  19995  Avatar  [Action, Adventure, Fantasy, Science Fiction]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [30]:
#lets check by this way also i.e of cast column row 0 new updated version of 3 cast name only
df["cast"][0]

['Sam Worthington', 'Zoe Saldana', 'Sigourney Weaver']

### Step-5:decorating the weired data available in the crew column accordingly to ours need
### HERE,BASICALLY WE WILL BE ONLY GRABBING THE DIRECTOR NAME FOR THIS IN LIST

In [31]:
#viewing what the data is like initially of any one of crew column
df.iloc[0].crew

# or df["crew"][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [32]:
def fetch_directname(value):
    L=[]
    for i in ast.literal_eval(value):
        if i["job"] == "Director":
            L.append(i["name"])
            break
    return L

In [33]:
#applying that helper function to our crew column which then updated to new one as we expected
df["crew"]=df["crew"].apply(fetch_directname)

In [34]:
df["crew"]

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4806, dtype: object

In [35]:
df.head(1)

id   title                                         genres  \
0  19995  Avatar  [Action, Adventure, Fantasy, Science Fiction]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

### Step-5:Changing the overview column's string paragraph to list format inorder to concatenate with the other to make a single one since we need to be same datatype for the merge
### HERE,BASICALLY WE WILL BE CONVERTING THIS STRING FORM PARAGRAPH IN LIST

In [36]:
#viewing what the data is like initially of any one of crew column
df.iloc[0].overview

# or df["overview"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [37]:
#applying a simple lambda function(for splitting string each words to list from) to our overview column which then updated to new one as we expected
df["overview"]=df["overview"].apply(lambda x:x.split())

In [38]:
df["overview"][0]

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.']

In [39]:
df.head(1)

id   title                                         genres  \
0  19995  Avatar  [Action, Adventure, Fantasy, Science Fiction]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

### Step-6:Before going for the concatination process(merging), we first need to remove all the spaces between the each words in the list of any columns 

In [40]:
# necessity to remove the spaces between the words of all the columns inorder to avoid being treated a single word as a 2 tags although in the same 
# for example:sam wilson--->samwilson
df["genres"]=df["genres"].apply(lambda x:[i.replace(" ","") for i in x])
df["keywords"]=df["keywords"].apply(lambda x:[i.replace(" ","") for i in x])
df["cast"]=df["cast"].apply(lambda x:[i.replace(" ","") for i in x])
df["crew"]=df["crew"].apply(lambda x:[i.replace(" ","") for i in x])

In [41]:
df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]  
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]  
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]  
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]

### Step-7:now,just concatination of the overview,keywords,genres,crew,cast columns into a single new column called "tagline"

In [42]:
df["tagline"]=df["overview"]+df["genres"]+df["keywords"]+df["cast"]+df["crew"]

In [43]:
df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   
2                    [Action, Adventure, Crime]   
3              [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, ScienceFiction]   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   
2  [A, cryptic, message, from, Bond’s, past, send...   
3  [Following, the, death, of, District, Attorney...   
4  [John, Carter, is, a, war-weary,, former, mili...   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   
2  [spy, basedonnovel, secretagent, sequel, mi6, ...   
3  [dccomics, crimefighter, terrorist, secretiden...   
4  [basedonnovel, mars, medallion, spacetravel, p...   

                                            cast                crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]      [JamesCameron]   
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]     [GoreVerbinski]   
2      [DanielCraig, ChristophWaltz, LéaSeydoux]         [SamMendes]   
3      [ChristianBale, MichaelCaine, GaryOldman]  [ChristopherNolan]   
4    [TaylorKitsch, LynnCollins, SamanthaMorton]     [AndrewStanton]   

                                             tagline  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [44]:
df["tagline"][0]

['In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron']

### Step-8:creating a new dataframe consisting only the requried columns(basically similar to dropping the unnecessary columns)

In [45]:
new_df=df[["id","title","tagline"]]

In [46]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                             tagline  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

### Step-9:converting the tagline column datas from the list to string again

In [47]:
new_df["tagline"]=new_df["tagline"].apply(lambda x:" ".join(x))

C:\Users\asus\AppData\Local\Temp\ipykernel_22040\2615836574.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tagline"]=new_df["tagline"].apply(lambda x:" ".join(x))


In [48]:
# checking
new_df["tagline"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [49]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                             tagline  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

### Step-10:converting the tagline column datas to lower case

In [50]:
new_df["tagline"]=new_df["tagline"].apply(lambda x:x.lower())

C:\Users\asus\AppData\Local\Temp\ipykernel_22040\1419782930.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tagline"]=new_df["tagline"].apply(lambda x:x.lower())


In [51]:
# checking
new_df["tagline"][1]

"captain barbossa, long believed to be dead, has come back to life and is headed to the edge of the earth with will turner and elizabeth swann. but nothing is quite as it seems. adventure fantasy action ocean drugabuse exoticisland eastindiatradingcompany loveofone'slife traitor shipwreck strongwoman ship alliance calypso afterlife fighter pirate swashbuckler aftercreditsstinger johnnydepp orlandobloom keiraknightley goreverbinski"

In [52]:
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                             tagline  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...  
2  a cryptic message from bond’s past sends him o...  
3  following the death of district attorney harve...  
4  john carter is a war-weary, former military ca...

### Additional step:using nltk library to perform stemming as later on we faced the problem of getting the similar words using countvectorizer

#### 1)importing nltk and it's required module and loading it

In [53]:
import nltk

In [54]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

#### 2)we will be using ps.stem so Let's see what it does first through example

In [55]:
#example
["dance","dancing","danced","dances"]

['dance', 'dancing', 'danced', 'dances']

In [56]:
ps.stem("dance")

'danc'

In [57]:
ps.stem("dancing")

'danc'

In [58]:
ps.stem("danced")

'danc'

In [59]:
ps.stem("dances")

'danc'

#### 3)Creating a def function where we first change our string pragraph of our tagline column to list format and then perfoming on each word the stemming by iteration 

In [60]:
def stemmed(my_text):
    y=[]
    for i in my_text.split():
        y.append(ps.stem(i))

    return " ".join(y)

In [61]:
#checking the working of our stemmed function(using the row 1 tagline column data here to check)
stemmed("captain barbossa, long believed to be dead, has come back to life and is headed to the edge of the earth with will turner and elizabeth swann. but nothing is quite as it seems. adventure fantasy action ocean drugabuse exoticisland eastindiatradingcompany loveofone'slife traitor shipwreck strongwoman ship alliance calypso afterlife fighter pirate swashbuckler aftercreditsstinger johnnydepp orlandobloom keiraknightley goreverbinski")

"captain barbossa, long believ to be dead, ha come back to life and is head to the edg of the earth with will turner and elizabeth swann. but noth is quit as it seems. adventur fantasi action ocean drugabus exoticisland eastindiatradingcompani loveofone'slif traitor shipwreck strongwoman ship allianc calypso afterlif fighter pirat swashbuckl aftercreditssting johnnydepp orlandobloom keiraknightley goreverbinski"

#### 4)using apply function to apply the above function to our tagline column(i.e updating with the root words in our tagline column)

In [62]:
new_df["tagline"]=new_df["tagline"].apply(stemmed)

C:\Users\asus\AppData\Local\Temp\ipykernel_22040\3287471488.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df["tagline"]=new_df["tagline"].apply(stemmed)


In [63]:
#checking the new tagline column by printing
new_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                             tagline  
0  in the 22nd century, a parapleg marin is dispa...  
1  captain barbossa, long believ to be dead, ha c...  
2  a cryptic messag from bond’ past send him on a...  
3  follow the death of district attorney harvey d...  
4  john carter is a war-weary, former militari ca...

In [64]:
#checking only tagline column 0th row data
new_df["tagline"][0]

'in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d samworthington zoesaldana sigourneyweav jamescameron'

#
# VECTORIZATION PART

### Step-1:Importing and loading sklearn library to perform vectorization

In [65]:
# Importing and loading 
from sklearn.feature_extraction.text import CountVectorizer    #importing countvectorizer from feature extraction.text of sklearn(API)
cv=CountVectorizer(max_features=5000, stop_words="english")    #loading it with adding parameters

### Step-2:using fit_transform to train to learn vocabulary and tranform it to count vectors(bag of words model)

In [66]:
# It converts text data (from the "tagline" column of your new_df DataFrame) into a numerical matrix using CountVectorizer along with explicitely 
# converting it to numpy array 
vectors=cv.fit_transform(new_df["tagline"]).toarray()

In [67]:
#printing the output of all movies
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], shape=(4806, 5000))

In [68]:
#printing the vector of only 2nd movie
vectors[2]

array([0, 0, 0, ..., 0, 0, 0], shape=(5000,))

### Step-3(optional):printing out the 5000 word vectors that we got(i.e top high 5000 frequency words)

In [69]:
#length checking(how many words are there)
len(cv.get_feature_names_out())

5000

In [70]:
#printing out all 5000 words
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      shape=(5000,), dtype=object)

In [71]:
#only printing out last 30 words
cv.get_feature_names_out()[-30:]

array(['xenophobia', 'yacht', 'yakuza', 'yard', 'year', 'yearn', 'years',
       'yellow', 'york', 'yorker', 'young', 'youngadult', 'youngboy',
       'younger', 'youngest', 'youngster', 'youth', 'yuppi', 'zacefron',
       'zachgalifianaki', 'zealand', 'zebra', 'zeu', 'zoe', 'zoesaldana',
       'zombi', 'zombies', 'zone', 'zoo', 'zooeydeschanel'], dtype=object)

# DISTANCE CALCULATION OF ALL MOVIES I.E EACH MOVIE WITH EACH MOVIE USING COSINE DISTANCE(I.E BASED ON ANGULAR DEGREE)

In [72]:
#importing an already made function from sklearn to perform cosine distance calculation
from sklearn.metrics.pairwise import  cosine_similarity

In [74]:
#placing our vector in this function to get result
similarity_check=cosine_similarity(vectors)    #placing our vectors variable having the bag of words model trained data

In [75]:
#checking the shape of it
similarity_check.shape

(4806, 4806)

###### got output:(4806,4806) as each movie is basically measuring distance with each movie available
###### i.e movie1 with movie1,movie1 with movie2,..........,movie1 with movie4806,movie2 with movie1,movie2 with movie2,movie2 with movie3,......,movie4806 with movie1,movie4806 with movie2,..........,movie4806 with movie4806)


In [77]:
#printing the output
similarity_check

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]], shape=(4806, 4806))

In [78]:
#let's understand what output of this array of array means

#Let's check for movie1
similarity_check[0]

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ], shape=(4806,))

##### got output:array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,0.        ], shape=(4806,))
##### it means for movie 1
##### where,
##### movie1 with movie1 = 1 similarity value
##### movie1 with movie2 = 0.08346223 similarity value

In [79]:
#let's check for movie2
similarity_check[1]

array([0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
       0.02615329], shape=(4806,))